File: src/test/scala/com/exasol/parquetio/RowParquetReaderTest.scala
package com.exasol.parquetio

import java.lang.Integer

import com.exasol.parquetio.data.GenericRow

import org.apache.parquet.example.data.simple.SimpleGroup
import org.apache.parquet.schema.MessageTypeParser

// [utest->dsn~read-parquet-file-contents~1]
class RowParquetReaderTest extends BaseParquetReaderTest {

  test("reads file contents") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required int32 col_int;
         |}
         |""".stripMargin
    )
    val input = Seq(1, 173, Int.MaxValue, Int.MinValue)
    withResource(getParquetWriter(schema, true)) { writer =>
      input.foreach { value =>
        val record = new SimpleGroup(schema)
        record.append("col_int", value)
        writer.write(record)
      }
    }
    val records = getRecords()
    val expected = input.map(v => GenericRow.of(Integer.valueOf(v)))
    assert(records.size === 4)
    assert(records === expected)
  }

  test("read file row values using positional index and field names") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required int32 col_int;
         |  required binary col_str;
         |}
         |""".stripMargin
    )
    val values = Seq(1, 10, 13)
    withResource(getParquetWriter(schema, true)) { writer =>
      values.foreach { value =>
        val record = new SimpleGroup(schema)
        record.append("col_int", value)
        record.append("col_str", s"$value")
        writer.write(record)
      }
    }
    val records = getRecords()
    values.zipWithIndex.foreach { case (value, index) =>
      assert(records(index).getValue(0) === value)
      assert(records(index).getValue(1) === s"$value")
      assert(records(index).getValue("col_int") === value)
      assert(records(index).getValue("col_str") === s"$value")
    }
  }

}

File: src/test/java/com/exasol/parquetio/RowParquetReaderTest.java
package com.exasol.parquetio;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import static org.junit.jupiter.api.Assertions.assertAll;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import org.apache.parquet.example.data.simple.SimpleGroup;
import org.junit.jupiter.api.Test;

import com.exasol.parquetio.data.GenericRow;
import com.exasol.parquetio.data.Row;

// [utest->dsn~read-parquet-file-contents~1]
class RowParquetReaderTest extends BaseParquetReaderTest {
    @Test
    void readsFileContents() throws IOException {
        final var schema = parseSchema("message test {", "  required int32 col_int;", "}");
        final List<Integer> input = List.of(1, 173, Integer.MAX_VALUE, Integer.MIN_VALUE);
        try (var writer = getParquetWriter(schema, true)) {
            for (final int value : input) {
                final var parquetRow = new SimpleGroup(schema);
                parquetRow.append("col_int", value);
                writer.write(parquetRow);
            }
        }
        final List<Row> records = getRecords();
        final List<GenericRow> expected = new ArrayList<>();
        for (final int value : input) {
            expected.add(GenericRow.of(value));
        }
        assertAll(() -> assertThat(records, hasSize(4)), () -> assertThat(records, equalTo(expected)));
    }

    @Test
    void readsFileRowValuesUsingPositionalIndexAndFieldNames() throws IOException {
        final var schema = parseSchema("message test {", "  required int32 col_int;", "  required binary col_str;",
                "}");
        final List<Integer> values = List.of(1, 10, 13);
        try (var writer = getParquetWriter(schema, true)) {
            for (final int value : values) {
                final var parquetRow = new SimpleGroup(schema);
                parquetRow.append("col_int", value);
                parquetRow.append("col_str", Integer.toString(value));
                writer.write(parquetRow);
            }
        }
        final List<Row> records = getRecords();
        for (int index = 0; index < values.size(); index++) {
            final int value = values.get(index);
            final Row row = records.get(index);
            assertAll(() -> assertThat(row.getValue(0), equalTo(value)),
                    () -> assertThat(row.getValue(1), equalTo(Integer.toString(value))),
                    () -> assertThat(row.getValue("col_int"), equalTo(value)),
                    () -> assertThat(row.getValue("col_str"), equalTo(Integer.toString(value))));
        }
    }
}
